### PR TITLE
feat: Add Zigzag Static Expert Load Balancing Placement Strategy for MoE Models

### PR DESCRIPTION
**Pull request by zkkython**
## Description:
This PR introduces a novel static expert load balancing placement strategy, called Zigzag, specifically designed for Mixture-of-Experts (MoE) models with multiple expert groups (such as the DeepSeek series) within SGLang. The strategy optimizes the distribution of experts across available devices (e.g., GPUs in a tensor-parallel setup) by implementing a staggered placement pattern. This aims to achieve better load balancing across expert parallel groups, which is particularly beneficial for models using grouped top-k routing, where experts are organized into logical groups and routing decisions are made within these groups 

## Motivation
Through analysis of multi-expert-group MoE models like DeepSeek, it was observed that experts within the same group tend to be selected together during inference. Concentrating these related experts on the same device can lead to load imbalance. The Zigzag strategy addresses this by intentionally distributing experts from the same logical group across different physical devices. This ensures that when the routing mechanism selects multiple experts from the same group, the computational load is naturally distributed, preventing individual devices from becoming bottlenecks and improving overall system throughput 

## Additional Notes:
Compatibility: This implementation is currently validated for MoE models with grouped experts, like DeepSeek. Its effectiveness may vary based on the model's specific expert routing architecture.
Future Work: As SGLang's deterministic inference support expands to include MoE models 
, the interaction between deterministic sampling and expert placement can be further explored. Optimizing the strategy for different parallel configurations (e.g., larger tensor parallelism sizes) is also a potential area for future improvement.
Reference: This implementation is inspired by the concept and results demonstrated in the corresponding vLLM PR #23745
## Modifications

Add Zigzag Static Expert Load Balancing Placement Strategy for MoE Models

## Accuracy Tests

Dataset	Accuracy Baseline	     Accuracy (zigzag-MR)
Gpqa	             68.18%                           68.18%
GSM8K                 95%                                94.84%

## Benchmarking and Profiling

Test Platform:

SGLang version: sgl-project/sglang:v0.5.5.post3
Model: DeepSeek-R1-FP8
GPU: H20(144G) x 8
Parallel config 1: tp=8, ep=8
Benchmark config: input_len=1024, output_len=128, request_rate=8, max_concurrency=8, num_prompts=32:

Service config:
```
python3 -m sglang.launch_server \
    --trust-remote-code \
    --init-expert-location round_robin \
    --cuda-graph-max-bs 12 \
    --cuda-graph-bs 1 2 3 4 5 6 7 8 9 10 11 12 \
    --disable-custom-all-reduce \
    --disable-radix-cache \
    --ep-size 8 \
    --moe-a2a-backend deepep \
    --model-path ${MODEL_PATH} \
    --port 30000 \
    --mem-fraction-static 0.87 \
    --tp-size 8  \
    --trust-remote-code \
    --host 0.0.0.0
```
Benchmark config:
```
python3 -m sglang.bench_serving \
    --backend sglang \
    --dataset-name random \
    --model ${MODEL_PATH} \
    --random-input-len 1024 \
    --random-output-len 128 \
    --random-range-ratio 0.5 \
    --dataset-path ./ShareGPT_V3_unfiltered_cleaned_split.json \
    --request-rate 8 \
    --max-concurrency 8 \
    --num-prompts 32 \
    --base-url http://127.0.0.1:30000 \
    --port 30001
```
ep_zigzag.log
image
default_ep.log
image

Conclusion: With only expert parallelism enabled, Zigzag improves throughput and end-to-end latency by approximately 2%.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/eplb/expert_location.py [hunk 1]
```diff
@@ -111,6 +111,48 @@ def init_trivial(
...
+                for j in range(num_local_physical_experts)
+            ]
+
+            rr_physical_to_logical_map = (
+                torch.tensor(
+                    physical_expert_indices_ordered_by_ep_group, dtype=torch.int32
+                ).repeat(num_layers, 1)
+                % num_logical_experts
+            )
+            return rr_physical_to_logical_map
```
[the remark this answers is not in the document]
Comment (zkkython) [reply python/sglang/srt/eplb/expert_location.py:146]: The design logic of the original nested loops for i in range(ep_size) for j in range(num_local_physical_experts) is as follows: iterate first by the EP Group dimension (ep_size), then by the physical experts within each group. Its purpose is to distribute replicas of the same logical expert evenly across different EP Groups, rather than "directly mapping by physical expert indices."
The reviewer’s modification (torch.arange(0, num_physical_experts)) only focuses on the one-dimensional mapping of "physical expert → logical expert," completely ignoring the core topological dimension of EP Groups. This degrades load balancing from "cross-Group balance" to "global one-dimensional balance," which is inconsistent with the expert parallelism design for business scenarios.

Original Code Logic:
physical_expert_indices_ordered_by_ep_group = [i + j * ep_size for i in range(ep_size) for j in range(num_local_physical_experts)]
Assuming ep_size=4 (4 EP Groups) and num_local_physical_experts=2 (2 local physical experts per group), the generated physical expert indices are: [0,4,1,5,2,6,3,7] (contrary to the reviewer’s assumption of "0,2,0,2"). The core of this sequence is: experts from each EP Group (i) are iterated first, followed by the local experts within the group (j). This ensures logical experts are assigned to different EP Groups in a round-robin fashion, preventing overload of a single Group.

Reviewer's Modification:
torch.arange(0, num_physical_experts) generates [0,1,2,3,4,5,6,7]. After mapping via % ep_size, logical experts 0-3 are assigned to Groups 0-3, but logical experts 4-7 are mapped back to Groups 0-3. While this appears "uniform," it ignores the constraint of "number of local experts per group," leading to uneven load distribution among physical experts within the same Group (e.g., physical experts 0 and 4 in Group 0 end up handling multiple logical experts simultaneously).
